File: 2021/05/vents.ts
import * as fs from 'fs';
import _ from 'lodash';

type Point = {
    x: number
    y: number
}

type Line = {
    from: Point
    to: Point
}

function isParallelToAxes(line: Line): boolean {
    return line.from.x == line.to.x || line.from.y == line.to.y
}

function axis(line: Line, d: 'x' | 'y'): number[] {
    const o = d == 'x' ? 'y' : 'x'

    if (line.from[d] == line.to[d]) {
        return Array(Math.abs(line.from[o] - line.to[o]) + 1).fill(line.from[d])
    } else {
        return _.range(line.from[d], line.to[d] + ((line.from[d] - line.to[d]) < 0 ? 1 : -1))
    }
}

function walk(line: Line, fn: (y: number, x: number) => void): void {
    const coordinates = _.zip(axis(line, 'y'), axis(line, 'x')) as unknown as [number, number][]

    coordinates.forEach(c => fn(c[0], c[1]))
}

function mapVents(lines: Line[]): number[][] {  
    const result: number[][] = []

    lines.forEach(line => {
        walk(line, (y, x) => {
            result[y] ||= []
            result[y][x] = (result[y][x] ||0) + 1
        })
    })
    
    return result
}

function overlapping(map: number[][]): number {
    return map.flat().filter(e => e > 1).length
}

const input = fs.readFileSync('input/05.txt').toString().trim().split('\n')
const lines: Line[] = input.map(line => {
    const [fromX, fromY, toX, toY] = line.split('->').flatMap(x => x.split(',')).map(x => parseInt(x))
    
    return {from: {x: fromX, y: fromY}, to: {x: toX, y: toY}}    
})

console.log(overlapping(mapVents(lines.filter(l => isParallelToAxes(l)))))
console.log(overlapping(mapVents(lines)))

File: 2021/05/vents.rb
Point = Struct.new :x, :y
Line = Struct.new :from, :to do
  def walk
    x_axis =
      if from.x == to.x
        Array.new (from.y - to.y).abs + 1, from.x
      elsif from.x < to.x
        from.x.upto to.x
      else
        from.x.downto to.x
      end

    y_axis =
      if from.y == to.y
        Array.new (from.x - to.x).abs + 1, from.y
      elsif from.y < to.y
        from.y.upto to.y
      else
        from.y.downto to.y
      end

    x_axis.zip(y_axis).each do |x, y|
      yield x, y
    end
  end

  def parallel_to_axes?
    from.x == to.x || from.y == to.y
  end
end

def read_input
  File.read('../input/05.txt').split("\n").map do |line|
    from_x, from_y, to_x, to_y = line.split('->').flat_map { _1.split(',').map(&:to_i) }

    Line.new Point.new(from_x, from_y), Point.new(to_x, to_y)
  end
end

def map_vents(lines)
  lines.each_with_object([]) do |line, map|
    line.walk do |x, y|
      map[y] ||= []
      map[y][x] = (map[y][x] || 0) + 1
    end
  end
end

def count_overlapping(map)
  map.flatten.compact.count { _1 > 1 }
end

def humanize(map)
  map.compact.map { |line| line.map { _1 ? _1 : '.' }.join }
end

input = read_input

p count_overlapping map_vents input.select(&:parallel_to_axes?)
p count_overlapping map_vents input
